src/containers/ImportResources/ImportResources.js

```diff
@@ -0,0 +1,221 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import React, { Component } from 'react';
+
+// import Link from 'react-router-dom';
+import {
+  ToastNotification,
+  TextInput,
+  FormLabel,
+  Button,
+  Tooltip
+} from 'carbon-components-react';
+import { connect } from 'react-redux';
+
+import '../../components/Definitions/Definitions.scss';
+import './ImportResources.scss';
+
+import { createPipelineRun } from '../../api';
+import ServiceAccountsDropdown from '../ServiceAccountsDropdown';
+import { getSelectedNamespace } from '../../reducers';
+
+export class ImportResources extends Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      repositoryURL: '',
+      serviceAccount: '',
+      submitSuccess: false,
+      invalidInput: false,
+      logsURL: '',
+      directory: ''
+    };
+  }
+
+  handleServiceAccount(data) {
+    this.setState({
+      serviceAccount: data.selectedItem.text
+    });
+  }
+
+  handleTextInput(event) {
+    const identifier = event.target.name;
+    const inputValue = event.target.value;
+    this.setState({
+      [identifier]: inputValue
+    });
+  }
+
+  handleSubmit(event) {
+    const { namespace } = this.props;
+    const pipelinename = 'pipeline0';
+    const gitresourcename = 'git-source';
+    const gitcommit = 'master';
+    const applydirectory = this.state.directory;
+    const repourl = this.state.repositoryURL;
+    const serviceaccount = this.state.serviceAccount;
+    const payload = {
+      pipelinename,
+      serviceaccount,
+      gitresourcename,
+      gitcommit,
+      repourl,
+      applydirectory
+    };
+
+    console.log(payload);
+
+    const promise = createPipelineRun(payload, namespace);
+    promise.then(headers => {
+      const logsURL = headers.get('Content-Location');
+      const trimmedLogsURL = logsURL.substring(logsURL.lastIndexOf('/') + 1);
+      const finalURL = '#/pipelines/pipeline0/runs/'.concat(trimmedLogsURL);
+      this.setState({
+        logsURL: finalURL
+      });
+      this.setState({
+        submitSuccess: true,
+        invalidInput: false
+      });
+    });
+    promise.catch(error => {
+      const statusCode = error.response.status;
+      switch (statusCode) {
+        case 500:
+          this.setState({
+            submitSuccess: false,
+            invalidInput: true
+          });
+          break;
+        default:
+      }
+    });
+    event.preventDefault();
+  }
+
+  render() {
+    this.handleServiceAccount = this.handleServiceAccount.bind(this);
```

Comment:
`bind` returns a new function, so you're replacing these on every render. A better place for this would be the constructor, however even that's not necessary. Instead you can bind the functions in-place by writing them as follows:

```js
class ImportResources extends Component {
  state = {
    ...
  };

  handleServiceAccount = data => {
    this.setState(...);
  }
}
```

Reply:
I made this change and the handle code is working all good however my tests are now failing with `Cannot spy the handleTextInput property because it is not a function
; undefined given instead` does my code look ok to you and how can I fix this please?

Reply:
Ideally the test wouldn't need to spy on this function and you could just assert on some user-visible change in the component, e.g. value in the input has updated.

If however this is not possible, you'll have to go back to the old method of binding:

1. in the constructor:
`this.handleTextInput = this.handleTextInput.bind(this);`

2. handleTextInput
```
  handleTextInput(event) {
    const identifier = event.target.name;
    const inputValue = event.target.value;
    this.setState({
      [identifier]: inputValue
    });
  }
```
3. in the test you can then `spyOn(ImportResource.prototype, 'handleTextInput')` as you have already

Reply:
I updated the test to check for purely the change in text on the screen but I can't figure out now why it is not working, please could you take a look?

Reply:
@AlanGreene I am getting `Unable to find an element with the text: /Invalid URL here/i.` even though the html that is printed after this failure seems to show that the value has been updated correctly for that field, shouldn't this mean that the `getByText` for this value should succeed?